docusaurus/docs/engineering/boba-backend/backend-development-outline.md

```diff
@@ -0,0 +1,109 @@
+---
+sidebar_position: 2
+---
+
+# Backend development overview
+
+While every feature is unique, developing on boba backend usually involves changes to one or more of these entities:
+
+- **API endpoints:** the actions the backend makes available to clients.
+- **Permissions:** the flags that can be associated to a role to allow users to take specific actions.
+- **Database:** the tables and queries used to store and retrieve data from permanent storage.
+- **Tests:** the automated checks that ensure code runs (and continues to run) as expected.
+
+This page goes over the main steps involved in making changes to each of these. This is not the only order
+in which these changes can be tackled, and not all changes will involve all steps.
+
+## API endpoints development
+
+### Feature design phase
+
+At the end of the feature design phase, you should have the following information about any endpoint you need to add or modify:
+
+- **HTTP method and URL (required):** for example, `DELETE threads/:thread_id`
+
+- **HTTP request payload (optional):** the shape of the data sent by the client as part of the request body (most often with `POST` requests).
+
+  Example (from `POST posts/:post_id/contributions`):
+
+  ```ts
+  - content (required): string
+  - whisper_tags (optional): string[]
+  - index_tags (optional): string[]
+  - ...
+  ```
+
+- **HTTP response codes and payloads (required):** a list of the response codes returned by the endpoint.
+
+  Example (from `POST posts/:post_id/contributions`):
+
+  - **HTTP 200: Contribution created**
+
+    **Payload:** the contribution just created.
+
+    ```ts
+      - content: string
+      - whisper_tags: string[]
+      - index_tags: string[]
+      - ...
+    ```
+
+  - **HTTP 401: No authenticated user.**
+  - **HTTP 403: Unauthorized user.**
+  - **HTTP 404: Post not found.**
+
+### Development phase
+
+During the development phase you should:
+
+1. **Identify where the API endpoint should be placed:** see [Routes and Endpoints Structure](docs/engineering/boba-backend/APIs/creating-endpoints#structure)
+2. **Write the endpoint function** (if not already present): at this stage, it will simply throw [HTTP 501 Not Implemented](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501) error.
```

Comment:
```suggestion
2. **Write the endpoint function** (if not already present): at this stage, it will simply throw and [HTTP 501 Not Implemented](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501) error.
```